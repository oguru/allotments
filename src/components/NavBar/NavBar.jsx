import React, {
   useEffect,
   useRef,
   useState
} from "react";
import {CSSTransition} from "react-transition-group";
import NavLink from "../NavLink";
import PropTypes from "prop-types";
import styles from "./NavBar.module.scss";

const NavBar = (props) => {
   const {isActive} = props;

   NavBar.propTypes = {
      isActive: PropTypes.func
   };

   const [isNavOpen, setIsNavOpen] = useState(false);
   const [isLargeScreen, setIsLargeScreen] = useState(true);
   const [menuHeight, setMenuHeight] = useState("");

   useEffect(() => {
      const mediaQuery = window.matchMedia("(min-width: 768px)");

      handleMediaQueryChange(mediaQuery);
      mediaQuery.addEventListener("change", (mQuery) => handleMediaQueryChange(mQuery));

      return () => {
         mediaQuery.removeEventListener(handleMediaQueryChange);
      };
   }, []);

   const handleMediaQueryChange = mQuery => {
      if (mQuery.matches) {
         setIsLargeScreen(true);
      } else {
         setIsLargeScreen(false);
      }
   };

   const calcOpenNavHeight = () => {
      const height = navLinkGroup.current.offsetHeight;
      setMenuHeight(height);
   };

   const mobNavHeight =
    isLargeScreen ?
       "unset" : isNavOpen ?
          `${menuHeight + 16}px` : 0;

   const closeNav = () => {
      setIsNavOpen(false);
   };

   const navIconAnim = isNavOpen ? "navCross" : "";
   const navLinkGroup = useRef(null);

   return (
      <nav
         className={`
            ${styles.navMain}
            d-flex
            align-items-center`
         }
      >
         <div className="
            d-flex
            justify-content-between
            align-items-center
            container"
         >
            <a
               className={`
                  ${styles.navBrandText} 
                  navbar-brand
               `}
               data-test="navBrand"
               href="/"
            >
               Stechford Allotments
            </a>
            <div
               className={styles.burgerIcon}
               data-test="burgerIcon"
               onClick={() => setIsNavOpen(!isNavOpen)}
            >
               <span
                  className={styles[navIconAnim]}>
               </span>
               <span
                  className={styles[navIconAnim]}>
               </span>
               <span
                  className={styles[navIconAnim]}>
               </span>
            </div>
            <div
               className={styles.navOverlay}
               data-test="navOverlay"
               style={{height: `${mobNavHeight}`}}
            >
               <CSSTransition
                  classNames={{...styles}}
                  in={isNavOpen}
                  nodeRef={navLinkGroup}
                  onEnter={calcOpenNavHeight}
                  timeout={500}
               >
                  <div
                     className={`${styles.navBarLinkGroup}`}
                     data-test="navBarLinkGroup"
                     ref={navLinkGroup}
                  >
                     <NavLink
                        closeNav={() => closeNav()}
                        isActive={isActive("/")}
                        path="/"
                        linkText="Home"
                     />
                     <NavLink
                        closeNav={() => closeNav()}
                        isActive={isActive("/about")}
                        path="/about"
                        linkText="About"
                     />
                     <NavLink
                        closeNav={() => closeNav()}
                        isActive={isActive("/articles")}
                        path="/articles"
                        linkText="Articles"
                     />
                     <NavLink
                        closeNav={() => closeNav()}
                        isActive={isActive("/info")}
                        path="/info"
                        linkText="Info"
                     />
                  </div>
               </CSSTransition>
            </div>
         </div>
      </nav>
   );
};

export default NavBar;