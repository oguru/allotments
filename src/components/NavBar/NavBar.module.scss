@import "../../variables.scss";
@import "../../../node_modules/bootstrap/scss/mixins/breakpoints";
@import "../../../node_modules/bootstrap/scss/functions";
@import "../../../node_modules/bootstrap/scss/variables";

.navMain {
    background-color: $primary-color;
    border-bottom: 2px solid rgba(255, 255, 255, 0);
    height: 64px;
    position: relative;
    z-index: 1000;

    .burgerIcon {
        cursor: pointer;
        height: 22px;
        position: relative;
        top: 2px;
        width: 25px;
        
        span {
            background-image: url(../../images/burger-texture-01.png);
            background-size: 60px 3px;
            border-radius: $base-spacing;
            display: block;
            height: 3px;
            position: absolute;
            top: 0;
            transform: rotate(0deg) translate(0px);
            transition: 0.4s ease-in-out;
            width: 25px;
        }

        span:nth-child(2) {
            background-position-x: 33%;
            top: 9px;
        }

        span:nth-child(3) {
            background-position-x: 66%;
            top: 18px;
        }

        span.navCross:nth-child(1) {
            height: 3px;
            top: 7.5px;
            transform: rotate(135deg);
        }

        span.navCross:nth-child(2) {
            opacity: 0;
            transform: translateX(-100%);
        }

        span.navCross:nth-child(3) {
            height: 3px;
            top: 7.5px;
            transform: rotate(-135deg);
        }
    }

    .navBrandText {
        color: white;
        font-family: 'Fredericka the Great', cursive;
        font-size: 1.5rem;
        text-shadow: 0px 0px 0px rgb(180, 180, 180);
    }

    .navOverlay {
        @include nav-t-duration;
        background-color: $bg-color;
        left: 0;
        overflow: hidden;
        position: absolute;
        top: $base-spacing * 8;
        transition-property: height;
        width: 100vw;
    }

    .navBarLinkGroup {
        display: none;
        position: relative;
    }

    .enter {        
        display: block;
        transform: translateY(-100%);
    }
    
    .enterActive {
        @include nav-t-duration;
        opacity: 1;
        transform: translateY(0);
    }
    
    .enterDone {
        display: block;
    }
    
    .exit {
        @include nav-t-duration;
        display: block;
        opacity: 1;
        transform: translateY(0);
    }
    
    .exitActive {
        display: block;
        transform: translateY(-100%);
    }

    .exitDone {
        display: none;
    }
}

@include media-breakpoint-up(sm) {
    .navMain {
        .navBrandText {
            font-size: 2rem;
        }
    }
}

@include media-breakpoint-up(md) {
    .navMain {
        border: unset;

        .burgerIcon {
            display: none;
        }

        .enter,
        .exitActive {
            transform: translate(0);
        }
        
        .navBarLinkGroup {
            display: flex;
            justify-content: flex-end;
            position: unset;            
        }

        .navBrandText {
            font-size: 1.875rem;
        }
        
        .navOverlay {
            background: none;
            position: static;
        }
    }
}

@include media-breakpoint-up(lg) {
    .navMain {
        .navBrandText {
            font-size: 2rem;
        }
    }
}