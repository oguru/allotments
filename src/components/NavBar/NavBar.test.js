import {MemoryRouter} from "react-router-dom";
import {checkProps, findByTestAttr} from "../../../utils/utils";
import {mount, shallow} from "enzyme";
import {cleanup, render} from "@testing-library/react";
import MatchMediaMock from "jest-matchmedia-mock";
import React from "react";
import NavBar from "./NavBar";
import ReactDOM from "react-dom";
import {act} from "react-dom/test-utils";

describe(
   "NavBar tests", () => {
      let component;
      let matchMedia;
      const isActive = jest.fn();

      const fireResize = (width) => {
         window.innerWidth = width;
         window.dispatchEvent(new Event("resize"));
      };

      beforeEach(() => {
         matchMedia = new MatchMediaMock();
         component = mount(<MemoryRouter>
            <NavBar isActive={isActive} />
         </MemoryRouter>);
      });

      afterEach(() => {
         cleanup();
         matchMedia.clear();
      });

      it(
         "should render without errors", () => {
            expect(render(component)).toBeTruthy();
         }
      );

      it(
         "should match the snapshot", () => {
            expect(component).toMatchSnapshot();
         }
      );

      it(
         "should have a nav brand/title with the correct title", () => {
            expect(findByTestAttr(
               component, "navBrand"
            ).text()).toBe("Stechford Allotments");
         }
      );

      test(
         "nav brand/title should have the correct re-direct path to home", () => {
            expect(findByTestAttr(
               component, "navBrand"
            ).filter("[href='/']").length).toBe(1);
         }
      );

      it(
         "should contain 4 nav links", () => {
            expect(findByTestAttr(
               component, "navLink"
            ).length).toBe(4);
         }
      );

      it(
         "should have the correct re-direct path on each nav link", () => {
            const navLinks = findByTestAttr(
               component, "navLink"
            );

            expect(navLinks.find("[href='/about']").text()).toBe("About");

            expect(navLinks.find("[href='/']").text()).toBe("Home");

            expect(navLinks.find("[href='/info']").text()).toBe("Info");

            expect(navLinks.find("[href='/articles']").text()).toBe("Articles");
         }
      );

      it(
         "should have a burger menu icon", () => {
            expect(findByTestAttr(
               component, "burgerIcon"
            ).length).toBe(1);
         }
      );

      test(
         "burger menu icon elements should gain, then lose css class on click", () => {
            const burgerIcon = findByTestAttr(
               component, "burgerIcon"
            );

            expect(burgerIcon.children(".navCross").length).toBe(0);

            burgerIcon.simulate("click");

            expect(findByTestAttr(
               component, "burgerIcon"
            ).children(".navCross").length).toBe(3);

            burgerIcon.simulate("click");

            expect(findByTestAttr(
               component, "burgerIcon"
            ).children(".navCross").length).toBe(0);
         }
      );

      test(
         "NavBar PropTypes check should not throw a warning", () => {
            const expectedProps = {
               isActive: jest.fn()
            };
            const propsErr = checkProps(
               NavBar, expectedProps
            );
            expect(propsErr).toBeUndefined();
         }
      );

      // unworking test for media query
      // test(
      //    "temp test", () => {
      //       console.error(
      //          "pre-test text: ", findByTestAttr(
      //             component, "tempTest"
      //          ).text()
      //       );

      //       fireResize(200);
      //       component.setProps({});

      //       const mediaQuery = "(min-width: 768px)";
      //       const listener = jest.fn();
      //       const mql = window.matchMedia(mediaQuery);

      //       mql.addEventListener(
      //          "change", ev => {
      //             console.error(
      //                "event: ", ev
      //             );
      //             // ev.matches && listener()
      //          }
      //       );

      //       act(() => {
      //          matchMedia.useMediaQuery(mediaQuery);
      //       });

      //       console.error(findByTestAttr(
      //          component, "tempTest"
      //       ).text());

      //       // console.error(
      //       //    mql.matches, mql.media
      //       // );

      //       // fireResize(100);
      //       // component.setProps({});

      //       // console.error(findByTestAttr(
      //       //    component, "tempTest"
      //       // ).text());

      //       // console.error(window.innerWidth);

      //    }
      // );

      it(
         "should open, then close the mobile navbar when burger menu is clicked twice", () => {
            const navOverlay = findByTestAttr(
               component, "navOverlay"
            );

            const burgerIcon = findByTestAttr(
               component, "burgerIcon"
            );

            const closedHeight = navOverlay.instance().style.height;

            burgerIcon.simulate("click");

            expect(navOverlay.instance().style.height).not.toEqual(closedHeight);

            burgerIcon.simulate("click");

            expect(navOverlay.instance().style.height).toEqual(closedHeight);
         }
      );

      it(
         "should close the mobile navbar when a link is clicked", () => {
            const navOverlay = findByTestAttr(
               component, "navOverlay"
            );

            findByTestAttr(
               component, "burgerIcon"
            ).simulate("click");

            const openHeight = navOverlay.instance().style.height;

            findByTestAttr(
               component, "navLink"
            ).find("[href='/about']").simulate("click");

            expect(navOverlay.instance().style.height).not.toEqual(openHeight);

         }
      );

      // describe(
      //    "NavBar responsiveness tests", () => {

      //       it(
      //          "should show a burger icon on a mobile viewport only", () => {

      //             expect(findByTestAttr(
      //                component, "burgerIcon"
      //             ).length).toEqual(0);

      //             fireResize(700);

      //             expect(findByTestAttr(
      //                component, "burgerIcon"
      //             ).length).toEqual(1);
      //          }
      //       );
      //    }
      // );
   }
);