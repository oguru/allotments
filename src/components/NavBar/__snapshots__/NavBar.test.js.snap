// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar tests should match the snapshot 1`] = `ReactWrapper {}`;
